### ISSUE TITLE
When upgrading to 8.3, ensure that the runtime state gets migrated accordingly

### ISSUE THREAD

**Issue by romansmirnov** [camunda/zeebe#14252]
**Description**

With the introduction of multi-tenancy in the upcoming 8.3 release, the runtime state stored needs to be migrated because:

1. For most of the column families, the key layout changes, i.e., the respective `tenantId` becomes part of the key.
2. By including the `tenantId` in the key, the data of the corresponding column families needs to be reindexed to new column families.

Currently, the migration only happens when a Zeebe broker becomes the leader for a given partition after upgrading to a newer Zeebe version:

https://github.com/camunda/zeebe/blob/7598696d5af5eebbf233f64deecb6cbeff0f9355/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/StreamProcessor.java#L325-L338

(Please note: The `DbMigrationController` - to run the actual migration - is called by ` l.onRecovered(streamProcessorContext)`.)

Now with the upcoming changes introduced by multi-tenancy, a Zeebe broker upgraded from 8.2 to 8.3 won't be able to replay anything because it won't find the necessary data anymore in RocksDB. This is because the key layout changed and the newly introduced column families (which are empty at that point). Meaning, the data is not migrated before the replay, so the latest snapshot corresponds to the 8.2 RocksDB schema. But when applying the events from the logstream, it assumes that the runtime state is already in the 8.3 RocksDB schema.

In a nutshell:
* After upgrading to 8.3, the followers won't be able to replay any command as long as the runtime state is not migrated to 8.3.
* Even after an upgrade a Zeebe broker - running in version 8.3 - becomes leader, it won't be able to replay. Hence, the leader won't process any new incoming command.

---

required by #13315 
related to #7248

**Comment by romansmirnov** [comment]
## Solution Sketches

Together with @Zelldon, we brainstormed a few options on how to solve the issue.

### 1. Run a migration app as a dedicated `InitContainer`

Basically, when upgrading to 8.3, a migration app runs as an [`initContainer`](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/) which migrates the latest snapshots. Once finished, the actual Zeebe broker starts in version 8.3.

Some notes on this approach:
* When the migration app runs, it migrates the latest snapshot as necessary (defined by the migration plan).
* Once finished, the migration app replaces the latest snapshot with an 8.3 snapshot.
* When the actual Zeebe broker starts in version 8.3, it can replay the events (as a follower or as a leader).
  * It requires that the snapshot is always upgraded previously. There could be a check to verify that the snapshot is in the expected version already, if not, it will prevent starting that partition. The log statement should point the user to run the migration app.
* In a rolling upgrade scenario, processing can continue and is not blocked by a migration (assuming the quorum is still given).
* Snapshots:
  * A follower running on version 8.3 does not accept install requests from a leader running on version 8.2.
  * A follower running on version 8.2 still continues accepting an install request from a leader running on version 8.3.
    * After installing the 8.3 snapshot, the follower will fail to replay any events, i.e., it won't be able to access the required data in RocksDb.
  * Both scenarios are acceptable because these are intermediate states and will eventually resolved once (all) Zeebe brokers are running on version 8.3.
* `initContainer`
  * Helm Charts: Requires some changes there, so when a user upgrades to 8.3 the migration app runs accordingly.
  * Controller (SaaS): Requires some change there as well.
* In a self-managed case (without Kubernetes), the user needs to run the migration app manually (needs to be documented)
* Relates to #7248

### 2. According to the given RocksDB schema use the corresponding `Db**State` classes

Basically, it's a kind of code duplication to have 8.2-`Db**State`-compatible classes and 8.3-`Db**State`-compatible classes in the code base. Depending on the RocksDB schema version, the respective `Db**State` classes will be used.

Some notes:
* Basically, everything remains the same as of now, i.e., with 8.3, events can be replayed, and only when becoming a leader the data get migrated to 8.3.
* Only a Zeebe broker running on version 8.2 won't be able to replay anything when receiving an 8.3 snapshot. This is acceptable as this broker gets upgraded to 8.3 eventually.

**Comment by Zelldon** [comment]
## 👬🏼 Pairing 2️⃣ 

Today, we investigate whether our current approach would work and whether the migration is executed. 

TL;DR; Yes migration has been successfully, BUT we still struggling whether our current approach is better than the init container. Both have pros and cons.

### Details

With our latest changes, we were able to run the CI successfully and also locally the RollingUpdateTests went threw. 

We extracted the logs to verify whether the migration was executed:


```
2023-09-12 12:48:04.401 [Broker-1] [zb-actors-0] [ZeebePartition-1] INFO 
      io.camunda.zeebe.broker.system - Transition to FOLLOWER on term 1 - transitioning ZeebeDb
2023-09-12 12:48:04.469 [Broker-1] [zb-actors-0] [GatewayTopologyManager] DEBUG
      io.camunda.zeebe.gateway - Received metadata change from Broker 1, partitions {}, terms {} and health {1=UNHEALTHY}.
2023-09-12 12:48:04.470 [Broker-1] [zb-actors-0] [GatewayTopologyManager] DEBUG
      io.camunda.zeebe.gateway - Received metadata change from Broker 1, partitions {}, terms {} and health {1=UNHEALTHY}.
2023-09-12 12:48:04.957 [Broker-1] [zb-actors-1] [SnapshotStore-1] DEBUG
      io.camunda.zeebe.logstreams.snapshot - Opened database from '/usr/local/zeebe/data/raft-partition/partitions/1/runtime'.
2023-09-12 12:48:04.960 [Broker-1] [zb-actors-1] [ZeebePartition-1] INFO 
      io.camunda.zeebe.broker.system - Transition to FOLLOWER on term 1 - transitioning Migration
2023-09-12 12:48:05.608 [Broker-1] [zb-actors-1] [ZeebePartition-1] INFO 
      org.camunda.feel.FeelEngine - Engine created. [value-mapper: CompositeValueMapper(List(io.camunda.zeebe.feel.impl.MessagePackValueMapper@564d37c6)), function-provider: io.camunda.zeebe.feel.impl.FeelFunctionProvider@12dbb1c0, clock: io.camunda.zeebe.el.impl.ZeebeFeelEngineClock@56eaca83, configuration: Configuration(false)]
2023-09-12 12:48:06.025 [Broker-1] [zb-actors-1] [ZeebePartition-1] INFO 
      org.camunda.dmn.DmnEngine - DMN-Engine created. [value-mapper: CompositeValueMapper(List(io.camunda.zeebe.feel.impl.MessagePackValueMapper@768ef71e)), function-provider: org.camunda.feel.context.FunctionProvider$EmptyFunctionProvider$@2ca26c7d, audit-loggers: List(), configuration: Configuration(false,false,false)]
2023-09-12 12:48:06.027 [Broker-1] [zb-actors-1] [ZeebePartition-1] INFO 
      org.camunda.feel.FeelEngine - Engine created. [value-mapper: CompositeValueMapper(List(org.camunda.dmn.NoUnpackValueMapper@467ceed5)), function-provider: org.camunda.feel.context.FunctionProvider$EmptyFunctionProvider$@2ca26c7d, clock: SystemClock, configuration: Configuration(false)]
2023-09-12 12:48:06.218 [Broker-1] [zb-actors-1] [ZeebePartition-1] INFO 
      io.camunda.zeebe.engine.state.migration - Starting processing of migration tasks (use LogLevel.DEBUG for more details) ... 
2023-09-12 12:48:06.222 [Broker-1] [zb-actors-1] [ZeebePartition-1] DEBUG
      io.camunda.zeebe.engine.state.migration - Found 9 migration tasks: ProcessMessageSubscriptionSentTimeMigration, MessageSubscriptionSentTimeMigration, TemporaryVariableMigration, DecisionMigration, DecisionRequirementsMigration, ProcessInstanceByProcessDefinitionMigration, ProcessDefinitionVersionMigration, JobTimeoutCleanupMigration, JobBackoffCleanupMigration
2023-09-12 12:48:06.246 [Broker-1] [zb-actors-1] [ZeebePartition-1] DEBUG
      io.camunda.zeebe.engine.state.migration - Skipping ProcessMessageSubscriptionSentTimeMigration migration (1/9).  It was determined it does not need to run right now.
2023-09-12 12:48:06.249 [Broker-1] [zb-actors-1] [ZeebePartition-1] DEBUG
      io.camunda.zeebe.engine.state.migration - Skipping MessageSubscriptionSentTimeMigration migration (2/9).  It was determined it does not need to run right now.
2023-09-12 12:48:06.250 [Broker-1] [zb-actors-1] [ZeebePartition-1] DEBUG
      io.camunda.zeebe.engine.state.migration - Skipping TemporaryVariableMigration migration (3/9).  It was determined it does not need to run right now.
2023-09-12 12:48:06.253 [Broker-1] [zb-actors-1] [ZeebePartition-1] DEBUG
      io.camunda.zeebe.engine.state.migration - Skipping DecisionMigration migration (4/9).  It was determined it does not need to run right now.
2023-09-12 12:48:06.254 [Broker-1] [zb-actors-1] [ZeebePartition-1] DEBUG
      io.camunda.zeebe.engine.state.migration - Skipping DecisionRequirementsMigration migration (5/9).  It was determined it does not need to run right now.
2023-09-12 12:48:06.255 [Broker-1] [zb-actors-1] [ZeebePartition-1] DEBUG
      io.camunda.zeebe.engine.state.migration - Starting ProcessInstanceByProcessDefinitionMigration migration (6/9)
2023-09-12 12:48:06.257 [Broker-1] [zb-actors-1] [ZeebePartition-1] DEBUG
      io.camunda.zeebe.engine.state.migration - ProcessInstanceByProcessDefinitionMigration migration completed in 2 ms.
2023-09-12 12:48:06.258 [Broker-1] [zb-actors-1] [ZeebePartition-1] DEBUG
      io.camunda.zeebe.engine.state.migration - Finished ProcessInstanceByProcessDefinitionMigration migration (6/9)
2023-09-12 12:48:06.262 [Broker-1] [zb-actors-1] [ZeebePartition-1] DEBUG
      io.camunda.zeebe.engine.state.migration - Starting ProcessDefinitionVersionMigration migration (7/9)
2023-09-12 12:48:06.264 [Broker-1] [zb-actors-1] [ZeebePartition-1] DEBUG
      io.camunda.zeebe.engine.state.migration - ProcessDefinitionVersionMigration migration completed in 0 ms.
2023-09-12 12:48:06.264 [Broker-1] [zb-actors-1] [ZeebePartition-1] DEBUG
      io.camunda.zeebe.engine.state.migration - Finished ProcessDefinitionVersionMigration migration (7/9)
2023-09-12 12:48:06.266 [Broker-1] [zb-actors-1] [ZeebePartition-1] DEBUG
      io.camunda.zeebe.engine.state.migration - Starting JobTimeoutCleanupMigration migration (8/9)
2023-09-12 12:48:06.268 [Broker-1] [zb-actors-1] [ZeebePartition-1] DEBUG
      io.camunda.zeebe.engine.state.migration - JobTimeoutCleanupMigration migration completed in 1 ms.
2023-09-12 12:48:06.269 [Broker-1] [zb-actors-1] [ZeebePartition-1] DEBUG
      io.camunda.zeebe.engine.state.migration - Finished JobTimeoutCleanupMigration migration (8/9)
2023-09-12 12:48:06.269 [Broker-1] [zb-actors-1] [ZeebePartition-1] DEBUG
      io.camunda.zeebe.engine.state.migration - Starting JobBackoffCleanupMigration migration (9/9)
2023-09-12 12:48:06.270 [Broker-1] [zb-actors-1] [ZeebePartition-1] DEBUG
      io.camunda.zeebe.engine.state.migration - JobBackoffCleanupMigration migration completed in 0 ms.
2023-09-12 12:48:06.278 [Broker-1] [zb-actors-1] [ZeebePartition-1] DEBUG
      io.camunda.zeebe.engine.state.migration - Finished JobBackoffCleanupMigration migration (9/9)
2023-09-12 12:48:06.279 [Broker-1] [zb-actors-1] [ZeebePartition-1] INFO 
      io.camunda.zeebe.engine.state.migration - Completed processing of migration tasks (use LogLevel.DEBUG for more details) ... 
2023-09-12 12:48:06.280 [Broker-1] [zb-actors-1] [ZeebePartition-1] DEBUG
      io.camunda.zeebe.engine.state.migration - Executed 4 migration tasks: ProcessInstanceByProcessDefinitionMigration, ProcessDefinitionVersionMigration, JobTimeoutCleanupMigration, JobBackoffCleanupMigration
2023-09-12 12:48:06.281 [Broker-1] [zb-actors-1] [ZeebePartition-1] INFO 
      io.camunda.zeebe.broker.system - Transition to FOLLOWER on term 1 - transitioning QueryService
2023-09-12 12:48:06.286 [Broker-1] [zb-actors-1] [ZeebePartition-1] INFO 
      io.camunda.zeebe.broker.system - Transition to FOLLOWER on term 1 - transitioning BackupStore
2023-09-12 12:48:06.288 [Broker-1] [zb-actors-1] [ZeebePartition-1] INFO 
      io.camunda.zeebe.broker.system - Transition to FOLLOWER on term 1 - transitioning BackupManager
2023-09-12 12:48:06.299 [Broker-1] [zb-actors-1] [ZeebePartition-1] INFO 
      io.camunda.zeebe.broker.system - Transition to FOLLOWER on term 1 - transitioning InterPartitionCommandService
2023-09-12 12:48:06.300 [Broker-1] [zb-actors-1] [ZeebePartition-1] INFO 
      io.camunda.zeebe.broker.system - Transition to FOLLOWER on term 1 - transitioning StreamProcessor
```

### Rolling update
 
Furthermore, we started a benchmark with 8.2.12 and accumulated some state (~100-200 MB). We migrated to our newest version without issues. We were able to see that even if the follower tries to become candidate the migration is only executed just once (state is not recreated)


```
DEBUG 2023-09-12T13:06:51.247417222Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Partition role transitioning from CANDIDATE to FOLLOWER in term 4
DEBUG 2023-09-12T13:06:51.247728551Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Paused replay for partition 2
INFO 2023-09-12T13:06:51.247761456Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Prepare transition from FOLLOWER on term 4 to FOLLOWER
INFO 2023-09-12T13:06:51.248317603Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Prepare transition from FOLLOWER on term 4 to FOLLOWER - preparing Admin API
INFO 2023-09-12T13:06:51.248912244Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Prepare transition from FOLLOWER on term 4 to FOLLOWER - preparing BackupApiRequestHandler
INFO 2023-09-12T13:06:51.249188123Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Prepare transition from FOLLOWER on term 4 to FOLLOWER - preparing ExporterDirector
DEBUG 2023-09-12T13:06:51.249513696Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Closed exporter director 'Exporter-2'.
INFO 2023-09-12T13:06:51.251732741Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Prepare transition from FOLLOWER on term 4 to FOLLOWER - preparing SnapshotDirector
DEBUG 2023-09-12T13:06:51.252178926Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Discard job io.camunda.zeebe.broker.system.partitions.impl.AsyncSnapshotDirector$$Lambda$1955/0x00007f0a709c7310 QUEUED from fastLane of Actor SnapshotDirector-2.
INFO 2023-09-12T13:06:51.252599231Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Prepare transition from FOLLOWER on term 4 to FOLLOWER - preparing StreamProcessor
DEBUG 2023-09-12T13:06:51.253011108Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Closed stream processor controller StreamProcessor-2.
INFO 2023-09-12T13:06:51.253647071Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Prepare transition from FOLLOWER on term 4 to FOLLOWER - preparing InterPartitionCommandService
INFO 2023-09-12T13:06:51.253984624Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Prepare transition from FOLLOWER on term 4 to FOLLOWER - preparing BackupManager
INFO 2023-09-12T13:06:51.254314456Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Prepare transition from FOLLOWER on term 4 to FOLLOWER - preparing BackupStore
INFO 2023-09-12T13:06:51.254605721Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Prepare transition from FOLLOWER on term 4 to FOLLOWER - preparing QueryService
INFO 2023-09-12T13:06:51.254905759Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Prepare transition from FOLLOWER on term 4 to FOLLOWER - preparing Migration
INFO 2023-09-12T13:06:51.255159637Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Prepare transition from FOLLOWER on term 4 to FOLLOWER - preparing ZeebeDb
INFO 2023-09-12T13:06:51.255443546Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Prepare transition from FOLLOWER on term 4 to FOLLOWER - preparing LogStream
INFO 2023-09-12T13:06:51.255778914Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Close appender for log stream logstream-raft-partition-partition-2
INFO 2023-09-12T13:06:51.256060849Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] On closing logstream logstream-raft-partition-partition-2 close 1 readers
INFO 2023-09-12T13:06:51.256370914Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Prepare transition from FOLLOWER on term 4 to FOLLOWER - preparing LogStorage
INFO 2023-09-12T13:06:51.256690345Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Preparing transition from FOLLOWER on term 4 completed
INFO 2023-09-12T13:06:51.256969580Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Transition to FOLLOWER on term 4 starting
INFO 2023-09-12T13:06:51.257242079Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Transition to FOLLOWER on term 4 - transitioning LogStorage
INFO 2023-09-12T13:06:51.257571625Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Transition to FOLLOWER on term 4 - transitioning LogStream
INFO 2023-09-12T13:06:51.259099903Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Transition to FOLLOWER on term 4 - transitioning ZeebeDb
INFO 2023-09-12T13:06:51.259420461Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Transition to FOLLOWER on term 4 - transitioning Migration
INFO 2023-09-12T13:06:51.260004017Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Engine created. [value-mapper: CompositeValueMapper(List(io.camunda.zeebe.feel.impl.MessagePackValueMapper@5636f4b1)), function-provider: io.camunda.zeebe.feel.impl.FeelFunctionProvider@75967a21, clock: io.camunda.zeebe.el.impl.ZeebeFeelEngineClock@7ebeb0c7, configuration: Configuration(false)]
INFO 2023-09-12T13:06:51.261972632Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] DMN-Engine created. [value-mapper: CompositeValueMapper(List(io.camunda.zeebe.feel.impl.MessagePackValueMapper@1f5702c8)), function-provider: org.camunda.feel.context.FunctionProvider$EmptyFunctionProvider$@79717f9a, audit-loggers: List(), configuration: Configuration(false,false,false)]
INFO 2023-09-12T13:06:51.262358849Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Engine created. [value-mapper: CompositeValueMapper(List(org.camunda.dmn.NoUnpackValueMapper@27358b52)), function-provider: org.camunda.feel.context.FunctionProvider$EmptyFunctionProvider$@79717f9a, clock: SystemClock, configuration: Configuration(false)]
INFO 2023-09-12T13:06:51.263245201Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Starting processing of migration tasks (use LogLevel.DEBUG for more details) ...
DEBUG 2023-09-12T13:06:51.263566123Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Found 9 migration tasks: ProcessMessageSubscriptionSentTimeMigration, MessageSubscriptionSentTimeMigration, TemporaryVariableMigration, DecisionMigration, DecisionRequirementsMigration, ProcessInstanceByProcessDefinitionMigration, ProcessDefinitionVersionMigration, JobTimeoutCleanupMigration, JobBackoffCleanupMigration
DEBUG 2023-09-12T13:06:51.263924221Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Migration was executed before ProcessMessageSubscriptionSentTimeMigration migration (1/9). It does not need to run again.
DEBUG 2023-09-12T13:06:51.264226823Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Migration was executed before MessageSubscriptionSentTimeMigration migration (2/9). It does not need to run again.
DEBUG 2023-09-12T13:06:51.264494405Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Migration was executed before TemporaryVariableMigration migration (3/9). It does not need to run again.
DEBUG 2023-09-12T13:06:51.264790749Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Migration was executed before DecisionMigration migration (4/9). It does not need to run again.
DEBUG 2023-09-12T13:06:51.265110907Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Migration was executed before DecisionRequirementsMigration migration (5/9). It does not need to run again.
DEBUG 2023-09-12T13:06:51.265375939Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Migration was executed before ProcessInstanceByProcessDefinitionMigration migration (6/9). It does not need to run again.
DEBUG 2023-09-12T13:06:51.265807952Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Migration was executed before ProcessDefinitionVersionMigration migration (7/9). It does not need to run again.
DEBUG 2023-09-12T13:06:51.266096299Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Migration was executed before JobTimeoutCleanupMigration migration (8/9). It does not need to run again.
DEBUG 2023-09-12T13:06:51.266355838Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Migration was executed before JobBackoffCleanupMigration migration (9/9). It does not need to run again.
INFO 2023-09-12T13:06:51.266572157Z [jsonPayload.context.partitionId: 2] [resource.labels.containerName: zeebe] Completed processing of migration tasks (use LogLevel.DEBUG for more details) ...
```

Further logs you can find [here](https://console.cloud.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22zeebe-io%22%0Aresource.labels.location%3D%22europe-west1-b%22%0Aresource.labels.cluster_name%3D%22zeebe-cluster%22%0Aresource.labels.namespace_name%3D%22rc-migration-rolling-update%22%0Alabels.k8s-pod%2Fapp%3D%22camunda-platform%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Fcomponent%3D%22zeebe-broker%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Finstance%3D%22rc-migration-rolling-update%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Fmanaged-by%3D%22Helm%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Fname%3D%22zeebe%22%0Alabels.k8s-pod%2Fapp_kubernetes_io%2Fpart-of%3D%22camunda-platform%22%0Alabels.%22k8s-pod%2Fstatefulset_kubernetes_io%2Fpod-name%22%3D%22rc-migration-rolling-update-zeebe-2%22;pinnedLogId=2023-09-12T13:06:25.458470405Z%2F73jglskafng2uz9s;summaryFields=jsonPayload%252Fcontext%252FpartitionId:false:32:beginning;cursorTimestamp=2023-09-12T13:06:51.270053088Z;duration=PT1H?project=zeebe-io)


We realized that we still have the issue of becoming ready only after a role transition is complete.

You can see that the [BrokerHealtCheckService](https://github.com/camunda/zeebe/blob/main/broker/src/main/java/io/camunda/zeebe/broker/system/monitoring/BrokerHealthCheckService.java#L150) marks the broker as ready as soon we either are a follower or leader on all partitions. But this is not the leader or follower of the raft, it is the [PartitionTransition](https://github.com/camunda/zeebe/blob/main/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/PartitionStartupAndTransitionContextImpl.java#L174-L185) 


This means if the migration takes too long and we have hit the readiness probe timeout our pod might be restarted and we have to migrate again.

## Challenges

### Init containers

With the init container approach, we have the issue, of how to write/overwrite the snapshot. We could think of that writing a separate snapshot with a suffix might work, and we read that on restore (preferred over the other).

But, the biggest issue is actually that when we start we might join the cluster and can receive a snapshot with an old version, which we can't use and are not able to migrate anymore since the init container is already done.

### Migration in a transition step

With the transition step, we block the complete transition. We might be able to improve the performance, but still, the migration for multi-tenancy might touch the complete state which is enormous. This becomes even more a problem when we migrate a large state. 

Taking long in the migration means that we defer the start of the partition and become ready, which might cause it to be restarted if we hit the readiness timeout. 

We thought about whether could mark us ready already before, but here the problem is that we might accept commands that we can process in time. This leads to timeouts in clients, which might lead to retries and blow up the problem.

## Open questions

Do we accept that the migration might take long? That we might hit the readiness probe? This is also currently the case. If a user has a larger state he should be advised to remove the readniness probe or increase the timeout. Is this acceptable? 

Do we want to combine both approaches? Migration normally via init-container, but also add it as transition step to handle the case where we receive a snapshot? But this will lead again to our problem of hitting the readiness probe. Might be not a solution.

**Comment by Zelldon** [comment]
I asked the [following](https://camunda.slack.com/archives/C037RS2JHB8/p1694532168848989) in the team channel:

> Today paired with Roman again and we run into some issues again regarding readiness probe. > Just to check it doesn't make sense to mark the broker earlier ready or? Because commands might be written which we can't process in time then timeout happen for clients and retries etc right?
>
> But this is only the case for leaders would it somehow make sense to mark earlier ready if we know we are "just" a follower. :thinking_face:


Input from @deepthidevaki regarding the ready status:

> Currently, the broker is marked as ready when all partitions are ready. A partition is ready after
> 
> - RaftPartition is ready
> - ZeebePartition is installed
> 
> RaftPartition is ready when it's log has caught up with the leader. This is important
> 
> - To prevent unavailability during RollingUpdate. We make sure that the restarted follower can be part of the quorum before restarting the next one.
> - To handle dataloss scenarios (like in SWIFT). When the partition is ready, we know that it has caught up with the leader. So it is safe to restart other nodes.
> 
> As far as I know, readiness probe is not used for controlling traffic in our case because the gateways directly talk to the broker pods, not through the service. So even if it is ready or not, gateway will be able to send the request to the brokers.
> 
> > We thought about whether could mark us ready already before, but here the problem is that we might accept commands that we can process in time. This leads to timeouts in clients, which might lead to retries and blow up the problem.
> 
> This might be ok. Because the TopologyManager is notified only after leader is fully installed. So the gateway will not send the requests to this leader. If the leader installation only completes after migration, then all listerners are notified only after that. So the CommandAPI also won't accept requests for this partition. Please verify again.
